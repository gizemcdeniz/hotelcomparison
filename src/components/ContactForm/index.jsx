import React from "react";
import { Container } from "react-bootstrap";
import { useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import InputErrorMessage from "../InputErrorMessage";
import firebase from "../../firebaseConfig";
import { getErrorClass } from "../../utils/formHelpers";
import "./index.scss";

export default function ContactForm({ handleShow }) {
  const { register, handleSubmit, errors, reset } = useForm();
  const { t } = useTranslation();

  const onSubmit = (data) => {
    handleShow();
    const newMessage = firebase
      .firestore()
      .collection("contactUsMessages")
      .doc();
    newMessage.set({
      name: data.fullName,
      email: data.email,
      message: data.messageContent,
    });
    reset();
  };

  return (
    <Container fluid className="contactForm">
      <form onSubmit={handleSubmit(onSubmit)}>
        <div className="formInput">
          <label htmlFor="fullName">
            {t("contactUs.contactUsForm.name.label")}
            <input
              className={getErrorClass(errors.fullName)}
              type="text"
              placeholder={t("contactUs.contactUsForm.name.placeholder")}
              name="fullName"
              id="fullName"
              aria-label="Full Name"
              ref={register()}
            />
          </label>
        </div>

        <div className="formInput">
          <label htmlFor="email">
            {t("contactUs.contactUsForm.email.label")}
            <input
              className={getErrorClass(errors.email)}
              type="email"
              placeholder={t("contactUs.contactUsForm.email.placeholder")}
              name="email"
              id="email"
              aria-label="Email address"
              ref={register({
                required: {
                  value: true,
                  message: t("contactUs.contactUsForm.email.error.required"),
                },
                // TODO: intall joi and use it for email validation
              })}
            />
            <InputErrorMessage error={errors.email} />
          </label>
        </div>
        <button type="submit" className="submitBtn">
          {t("contactUs.contactUsForm.submitBtn")}
        </button>
      </form>
    </Container>
  );
}
