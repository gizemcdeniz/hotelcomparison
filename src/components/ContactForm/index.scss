@import "../../style/variables";
@import "../../style/buttons";
@import "../../style/forms";
@import "../../style/bootstrapImports";

.contactForm {
  font: inherit;

  /* Inputs Style */
  input[type="text"],
  input[type="email"],
  textarea {
    @include inputElement;
  }

  textarea {
    height: 8rem;
  }

  input.inputError,
  textarea.inputError {
    @include inputErrorBorder;
  }

  button.submitBtn {
    @include primaryBtn;
    background-color: $successColor;
    font-weight: 800;

    &:hover {
      background-color: $lightColor;
      color: $primaryColor;
    }
  }

  .formInput {
    margin-bottom: 1.4rem;
    position: relative;

    label {
      width: 100%;
      font-weight: 500;
    }
  }
}

@include media-breakpoint-down(sm) {
  .contactForm {
    button.submitBtn {
      width: 100%;
    }
  }
}
