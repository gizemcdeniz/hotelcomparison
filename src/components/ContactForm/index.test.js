import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import ContactForm from "./index";

describe("ContactForm", () => {
  it("renders correctly and matches the snapshot", () => {
    const tree = renderer.create(<ContactForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  beforeEach(() => {
    render(<ContactForm />);
  });

  it("should display required error when value is invalid", async () => {
    await fireEvent.submit(screen.getByRole("button"));
    expect(await screen.findAllByRole("alert")).toHaveLength(2);
  });

  it("should display min length error when message is short", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /Email address/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /message/i }), {
      target: {
        value: "pass",
      },
    });

    await fireEvent.submit(screen.getByRole("button"));

    expect(await screen.findAllByRole("alert")).toHaveLength(1);
    expect(screen.getByRole("textbox", { name: /Email address/i }).value).toBe(
      "test@mail.com"
    );
    expect(screen.getByRole("textbox", { name: /message/i }).value).toBe(
      "pass"
    );
  });

  it("should not display error when value is valid", async () => {
    fireEvent.input(screen.getByRole("textbox", { name: /Full Name/i }), {
      target: {
        value: "Name Nami",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /Email address/i }), {
      target: {
        value: "test@mail.com",
      },
    });

    fireEvent.input(screen.getByRole("textbox", { name: /message/i }), {
      target: {
        value: "Long message to be sent",
      },
    });

    const alertElement = screen.queryByRole("alert");
    expect(alertElement).toBeNull();
  });

  it("should empty input fields after submission", async () => {
    expect(screen.getByRole("textbox", { name: /Full Name/i }).value).toBe("");
    expect(screen.getByRole("textbox", { name: /Email address/i }).value).toBe(
      ""
    );
    expect(screen.getByRole("textbox", { name: /message/i }).value).toBe("");
  });
});
