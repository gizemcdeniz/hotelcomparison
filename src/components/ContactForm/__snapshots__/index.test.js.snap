// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm renders correctly and matches the snapshot 1`] = `
<div
  className="contactForm container-fluid"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="formInput"
    >
      <label
        htmlFor="fullName"
      >
        contactUs.contactUsForm.name.label
        <input
          aria-label="Full Name"
          id="fullName"
          name="fullName"
          placeholder="contactUs.contactUsForm.name.placeholder"
          type="text"
        />
      </label>
    </div>
    <div
      className="formInput"
    >
      <label
        htmlFor="email"
      >
        contactUs.contactUsForm.email.label
        <input
          aria-label="Email address"
          id="email"
          name="email"
          placeholder="contactUs.contactUsForm.email.placeholder"
          type="email"
        />
      </label>
    </div>
    <div
      className="formInput"
    >
      <label
        htmlFor="messageContent"
      >
        contactUs.contactUsForm.message.label
        <textarea
          aria-label="message"
          id="messageContent"
          name="messageContent"
          placeholder="contactUs.contactUsForm.message.placeholder"
        />
      </label>
    </div>
    <button
      className="submitBtn"
      type="submit"
    >
      contactUs.contactUsForm.submitBtn
    </button>
  </form>
</div>
`;
